lib/cache/cache.go

```diff
@@ -1771,16 +1853,37 @@ func (c *Cache) fetch(ctx context.Context, confirmedKinds map[resourceKind]types
 // the event will be emitted via the fanout.
 func (c *Cache) processEvent(ctx context.Context, event types.Event) error {
 	resourceKind := resourceKindFromResource(event.Resource)
-	collection, ok := c.collections.byKind[resourceKind]
-	if !ok {
+
+	legacyCollection, legacyFound := c.legacyCacheCollections.byKind[resourceKind]
+	handler, handlerFound := c.collections.byKind[resourceKind]
+
+	switch {
+	case !legacyFound && !handlerFound:
 		c.Logger.WarnContext(ctx, "Skipping unsupported event",
 			"event_kind", event.Resource.GetKind(),
 			"event_sub_kind", event.Resource.GetSubKind(),
 		)
 		return nil
-	}
-	if err := collection.processEvent(ctx, event); err != nil {
-		return trace.Wrap(err)
+	case legacyFound:
+		if err := legacyCollection.processEvent(ctx, event); err != nil {
+			return trace.Wrap(err)
+		}
+	case handlerFound:
+		switch event.Type {
+		case types.OpDelete:
+			if err := handler.onDelete(event.Resource); err != nil {
+				if !trace.IsNotFound(err) {
+					c.Logger.WarnContext(ctx, "Failed to delete resource", "error", err)
+					return trace.Wrap(err)
+				}
+			}
+		case types.OpPut:
+			if err := handler.onUpdate(event.Resource); err != nil {
+				return trace.Wrap(err)
+			}
+		default:
+			c.Logger.WarnContext(ctx, "Skipping unsupported event type", "event", event.Type)
```

Comment:
This should cause a hard error, btw. The typical example is if we introduced `OpTruncate` meaning "delete everything" and forgot to handle it in this watcher - there's no way to let the cache continue running on a stream after getting an unknown event.

Reply:
There's been enough changes that this is not just moving code around, there's no need to copy bugs from old code.